src/Butler/Auth/OIDC.hs

```diff
@@ -0,0 +1,206 @@
+                            (oidc oidcEnv)
+                            (manager oidcEnv)
+                            (encodeUtf8 oauthState)
+                            (encodeUtf8 oauthCode)
+                setSession (UserName tokens.idToken.sub)
+
+    guestCallbackRoute :: Servant.Handler AuthResp
+    guestCallbackRoute = setSession "guest"
+
+    setSession :: UserName -> Servant.Handler AuthResp
+    setSession username = do
+        session <- liftIO $ runProcessIO os process $ newSession sessions username
+        liftIO (SAS.acceptLogin cookieSettings jwtSettings session.sessionID) >>= \case
+            Just r -> do
+                let page = workspaceUrl mWorkspace
+                pure $ r (script_ $ "window.location.href = " <> showT page)
+            Nothing -> error "oops?!"
+
+    indexRoute :: AuthResult SessionID -> Servant.Handler (Html ())
+    indexRoute = \case
+        Authenticated sessionID -> do
+            mSession <- atomically (lookupSession sessions sessionID)
+            case mSession of
+                Just _ ->
+                    pure $ mkIndexHtml (websocketHtml (workspaceUrl mWorkspace) sessionID)
+                Nothing -> do
+                    liftIO $ runProcessIO os process $ logError "no more auth?" ["id" .= sessionID]
+                    pure "Unknown session"
+        _ -> throwError $ err303{errHeaders = [("Location", encodeUtf8 (workspaceUrl mWorkspace) <> "_guest_cb")]}
+
+    loginRoute :: OIDCEnv -> Servant.Handler NoContent
+    loginRoute oidcEnv = do
+        redirectLocation <- liftIO genOIDCURL
+        throwError $ err303{errHeaders = [("Location", redirectLocation)]}
+      where
+        genOIDCURL :: IO ByteString
+        genOIDCURL = do
+            loc <-
+                O.prepareAuthenticationRequestUrl
+                    (mkSessionStore oidcEnv Nothing Nothing)
+                    oidcEnv.oidc
+                    [O.openId]
+                    mempty
+            pure . B.pack $ show loc
+
+cookieSettings :: CookieSettings
+cookieSettings =
+    defaultCookieSettings
+        { cookieSameSite = SameSiteStrict
+        , cookieXsrfSetting = Nothing
+        , cookieExpires = Just (UTCTime (fromGregorian 2030 1 1) 0)
+        }
+
+oIDCAuthApp :: Sessions -> (Html () -> Html ()) -> ProcessIO AuthApplication
+oIDCAuthApp sessions mkIndexHtml = do
+    JwkStorage myKey <- fst <$> newProcessMemory "display-key.jwk" (JwkStorage <$> liftIO generateKey)
+    let jwtSettings = defaultJWTSettings myKey
+    let cfg = cookieSettings :. jwtSettings :. EmptyContext
+    env <- ask
+    Just client_id <- liftIO $ getEnv "OIDC_ID"
+    Just client_password <- liftIO $ getEnv "OIDC_PASSWORD"
+    oidcenv <-
+        liftIO . initOIDCEnv $
+            OIDCProviderConfig
+                "https://accounts.google.com"
+                (decodeUtf8 client_id)
+                (decodeUtf8 client_password)
```

Comment:
It might be better to pass the client_id and client_password as parameter to the 'oIDCAuthApp' and the 'publicOIDCDisplayApp', perhaps using newtype wrapper?

Reply:
Sure I'll propose a followup with this.